# <font color='6495ED'>Python Functions</font>

## <font color='6495ED'>*BONUS 5. Iterators and Generators</font>

В этом юните вы познакомитесь с двумя удобными инструментами разработчика на <font color='6495ED'>*Python*</font> — итераторами и генераторами.

*Они позволяют обрабатывать элементы из структур данных по одному без избыточного использования памяти. Эти объекты обязательно пригодятся вам как специалисту в <font color='6495ED'>Data Science</font> при обработке больших потоков входных данных, которые порой будет невозможно одновременно вместить в оперативной памяти компьютера.*

Например, генераторы — это незаменимый объект в компьютерном зрении. Предположим, вы пытаетесь научить нейронную сеть отличать слонов от бегемотов. У вас есть большой набор изображений (пусть это будет 10 тысяч штук) со слонами и бегемотами. Вы пытаетесь прочитать этот набор и положить его в переменную, чтобы в дальнейшем отправить его в сеть.

Увы, ничего не выйдет. Если у вас нет огромных запасов оперативной памяти, вы просто не сможете вместить такой набор данных и получите ошибку <font color='red'>*MemoryError*</font> (ошибка памяти).

Тут-то вам и пригодится использование генераторов. Они позволяют организовать чтение и выдачу данных порциями, не нагружая оперативную память.

**В этом юните вы научитесь:**

- получать итераторы из встроенных структур данных;
- использовать итераторы в циклах;
- создавать генераторы и использовать оператор <font color='6495ED'>*yield*</font>;
- использовать списочные сокращения.

### <font color='6495ED'>ИТЕРАТОРЫ</font>

В переводе с английского <font color='6495ED'>*iterator*</font> означает «перечислитель». В <font color='6495ED'>*Python*</font> это такая структура данных, которая позволяет выдавать объекты по одному, когда её об этом просит пользователь.

Объекты, из которых можно извлечь итератор, называются **итерируемыми объектами**. Примеры итерируемых объектов — список, кортеж, словарь, множество.

Изучение принципа работы итераторов — наш первый шаг к пониманию работы цикла <font color='6495ED'>*for*</font> с различными структурами данных, генераторов, а также работы специальных функций в <font color='6495ED'>*Python*</font>, созданных на их основе.

*Работа итератора очень похожа на автомат со сладостями без возможности новой заправки. Клиент подходит к автомату, нажимает на кнопку «Купить» — автомат выдает ему сладость. Клиент снова нажимает на кнопку — автомат выдаёт следующий по очереди товар. Так продолжается до тех пор, пока в автомате не закончатся сладости. Когда клиент попробует обратиться к пустому автомату, тот выдаст ошибку.*

Именно такой подход и реализуют итераторы: они выдают своё содержимое по одному элементу, а когда элементы заканчиваются, выдаётся ошибка.

Все встроенные структуры данных в <font color='6495ED'>*Python*</font>, которые поддерживают проход по своим элементам в цикле <font color='6495ED'>*for*</font> (списки, словари, множества, кортежи, строки), содержат в себе итераторы и являются итерируемыми объектами. Чтобы получить итератор из структуры, используется функция <font color='6495ED'>*iter()*</font>.

Давайте рассмотрим простой пример — создадим  список из трёх чисел и выведем его на экран:

In [1]:
new_list = [12, 14, 16]
print(new_list)
# Будет напечатано:
# [12, 14, 16]

[12, 14, 16]


С помощью <font color='6495ED'>*print()*</font> мы узнали содержимое списка <font color='6495ED'>*new_list*</font>. Пока что ничего нового нет. 

Теперь извлечём итератор из списка <font color='6495ED'>*new_list*</font> с помощью функции <font color='6495ED'>**iter()**</font> и занесём его в новую переменную <font color='6495ED'>*iter_list*</font>:

In [2]:
# Извлечём итератор из список new_list 
iter_list = iter(new_list) 
print(iter_list)
# Будет напечатано 
# list_iterator object at 0x0000025DAA2D3B50

Теперь вместо содержимого вызов функции <font color='6495ED'>*(iter_list)*</font> <font color='6495ED'>*print*</font> показал нам, что <font color='6495ED'>*iter_list*</font> — это объект с типом данных <font color='6495ED'>*list_iterator*</font>, и он находится в ячейке памяти <font color='6495ED'>*0x0000025DAA2D3B50*</font> (ваша ячейка может отличаться). 

**А что насчёт привычной для списка индексации? Можем ли мы можем получить элемент по индексу?**

In [3]:
# Попробуем получить значение по индексу 2
print(iter_list[2])
# Возникнет ошибка:
# TypeError: 'list_iterator' object is not subscriptable

TypeError: 'list_iterator' object is not subscriptable

Нам не удалось получить конкретный объект по индексу: возникла ошибка, которая сообщает, что объект типа <font color='6495ED'>*list_iterator*</font> не позволяет получать его элементы по индексам. Таким образом, можно сделать вывод, что итераторы не поддерживают индексацию.

**Что же теперь делать с этим итератором?**

Оказывается, механизм получения данных из итератора отличается от механизма для привычных нам структур.

К любому итератору можно применить функцию <font color='6495ED'>*next()*</font>, которая возвращает следующий элемент из итератора. Сделаем так четыре раза:

In [ ]:
print(next(iter_list))
print(next(iter_list))
print(next(iter_list))
print(next(iter_list))
# Будет напечатано:
# 12
# 14
# 16
# StopIteration:

12
14
16


StopIteration: 

В первые три раза всё было в порядке, а вот на четвёртый раз возникла ошибка <font color='red'>*StopIteration*</font>. Она означает, что элементы в итераторе закончились и получить из него больше ничего нельзя. При дальнейших попытках обратиться к следующим элементам мы продолжим получать ошибки.

Это связано с тем, что каждый новый вызов <font color='6495ED'>*next()*</font> уменьшает количество элементов в итераторе на 1, пока элементы не закончатся полностью.

Работу функции <font color='6495ED'>*next()*</font> можно представить в виде следующей схемы:

![](dst3-u1-md5.2_4_1.png)

Как видите, итератор является «одноразовым» объектом: как только элементы в итераторе заканчиваются, он становится по сути бесполезным. 

Давайте посмотрим, как перебирать элементы итератора, если мы не знаем заранее, сколько элементов он содержит.

Рассмотрим небольшую задачу: пусть в некотором веб приложении есть список с различными типами пользователей (администраторы, гости и другие), но сами типы и их количество нам не известны. По каким-то причинам нам захотелось вывести весь список типов пользователей через итератор нашего списка.

Можно попробовать сделать это в цикле <font color='6495ED'>*while*</font>:

In [ ]:
users = ['admin', 'guest', 'root', 'anonymous']
iter_users = iter(users)

# Создаём бесконечный цикл while
# !while True:
# !    # Выводим следующий объект из итератора
# !    print(next(iter_users)) 
# Будет напечатано:
# admin
# guest
# root
# anonymous
# ! StopIteration:

Однако из-за ошибки прекращается дальнейшая работа программы, что нежелательно. Чтобы программа не ломалась, достаточно добавить обработку исключения <font color='red'>*StopIteration*</font> на каждом шаге получения следующего элемента. Напомним, обработка исключений производится с помощью конструкции <font color='6495ED'>*try except*</font>:

In [ ]:
# Создаём бесконечный цикл while
while True:
    # Создаём блок обработки исключений
    try:
        # Выводим следующий объект из итератора
        print(next(iter_users))
   # Отлавливаем исключение StopIteration
    except StopIteration:
        # Когда возникает исключение, выводим фразу на экран
        print("User list is over!")
        # Завершаем цикл
        break
 
# Будет напечатано:
# admin
# guest
# root
# anonymous
# User list is over!

admin
guest
root
anonymous
User list is over!


Всё, как и в прошлом примере. Мы извлекли из списка users его итератор <font color='6495ED'>*iter_users*</font>. Далее мы использовали цикл <font color='6495ED'>*while*</font>, в теле которого выводили следующий в элемент из итератора с помощью <font color='6495ED'>*next()*</font>. Отличие состоит в том, что мы производим действия в блоке <font color='6495ED'>*try*</font>, что позволяет нам отловить ошибку <font color='6495ED'>*StopIteration*</font> в блоке <font color='6495ED'>*except*</font>. Таким образом, когда возникает данная ошибка, мы выводим на экран фразу <font color='red'>*"User list is over!"*</font> и завершаем цикл с помощью <font color='6495ED'>*break*</font>.

Интересный факт:  в коде ранее мы, сами того не понимая, написали цикл <font color='6495ED'>*for*</font>, не используя цикл <font color='6495ED'>*for*</font>! Но об этом чуть позже.

Очевидно, что <font color='6495ED'>*while*</font> — не самый подходящий цикл для итераторов. Сейчас он больше служил показательным примером, как можно обрабатывать исключение <font color='red'>*StopIteration*</font>.

Стандартным способом работы с итератором всё же является цикл <font color='6495ED'>*for*</font> в своём обычном обличии. Напомним, что цикл <font color='6495ED'>*for*</font> выполняется до тех пор, пока в последовательности (итераторе) не закончатся элементы. Это как раз то, что нам нужно! Код будет иметь вид:

In [ ]:
iter_users = iter(users)
# Создаём цикл for по объектам из итератора
for user in iter_users:
    # Выводим текущий элемент на экран
    print(user)
 
# Будет напечатано:
# admin
# guest
# root
# anonymous

admin
guest
root
anonymous


Как видите, работать с итераторами в цикле <font color='6495ED'>*for*</font> предельно просто — даже обрабатывать исключения не потребовалось. Более того, вы знаете, что не обязательно извлекать итератор из списка — цикл <font color='6495ED'>*for*</font> отлично работает и с самим списком.

Но на самом деле «под капотом» цикл <font color='6495ED'>*for*</font> при работе со списком извлекает из него итератор с помощью <font color='6495ED'>*iter()*</font> и последовательно вызывает функцию <font color='6495ED'>*next()*</font>, а её результат заносит в переменную цикла. Когда элементы закончатся, возникнет ошибка <font color='red'>*StopIteration*</font>. Но цикл <font color='6495ED'>*for*</font> отлавливает эту ошибку и не производит дальнейших действий.

Полная схема работы цикла <font color='6495ED'>*for*</font> представлена ниже:

![](dst3-u1-md5.2_4_2.png)

На самом деле вы уже пользовались циклом <font color='6495ED'>*for*</font> для работы с итераторами ранее. Например, итерируемым объектом является объект <font color='6495ED'>*enumerate*</font>. Напомним, он позволяет получать не только элемент из последовательности, но и его индекс (номер, начиная от 0).

In [ ]:
iter_users = iter(users)
# Создаём цикл по элементам и индексам списка users
for i, user in enumerate(users):
    # Выводим индекс и элемент на экран через двоеточие
    print(i, user, sep=': ')
# Будет напечатано:
# 0: admin
# 1: guest
# 2: root
# 3: anonymous

0: admin
1: guest
2: root
3: anonymous


Можно перебирать объекты из объекта <font color='6495ED'>*enumerate*</font> по одному, как это было сделано выше в цикле <font color='6495ED'>*for*</font>. При этом можно также собрать все элементы из итератора в список с помощью функции <font color='6495ED'>*list()*</font>:

In [ ]:
users = ['admin', 'guest', 'root', 'anonymous']

# Создаём итератор enumerate из списка users
enum_users = enumerate(users)
# Получаем список из объекта enumerate
enum_list = list(enum_users)
print(enum_list)
# Будет напечатано:
# [(0, 'admin'), (1, 'guest'), (2, 'root'), (3, 'anonymous')]

[(0, 'admin'), (1, 'guest'), (2, 'root'), (3, 'anonymous')]


Итак, мы сохранили все пары (номер, значение) в список <font color='6495ED'>*enum_list*</font>. Предположим, мы захотели продолжить пользоваться объектом <font color='6495ED'>*enum_users*</font>, все значения из которого только что, в коде выше, были сохранены в список <font color='6495ED'>*enum_list*</font>:

In [ ]:
# Создаём цикл по элементам итератора
for i, user in enum_users:
    print(i, user, sep=': ')
# Ничего не будет напечатано

Почему-то с помощью цикла <font color='6495ED'>*for*</font> ничего не было напечатано. Попробуем метод <font color='6495ED'>*next()*</font>:

In [ ]:
#! print(next(enum_users))
# Возникнет ошибка:
# StopIteration:
#!

Возникла ошибка итерации. Дело в том, что в итераторе закончились элементы в тот момент, как только мы получили из него список в примере выше, поэтому попытка получить из него ещё элементы вызвала ошибку.

Однако в цикле <font color='6495ED'>*for*</font> ошибка не возникла. Это связано с тем, что цикл <font color='6495ED'>*for*</font> не «ругается» на пустой итератор, а просто не совершает с ним никаких действий.

Давайте подведём промежуточный итог по итераторам: 

Итератор — это объект-перечислитель, который выдаёт следующий элемент из своих значений с помощью <font color='6495ED'>*next()*</font> либо выбрасывает исключение <font color='red'>*StopIteration*</font>, когда элементы заканчиваются.
Итерируемый объект — это объект, который может предоставить нам итератор. По его содержимому можно пройтись в цикле <font color='6495ED'>*for*</font>. Встроенные итерируемые объекты — список, кортеж, строка, словарь, множество
Все встроенные итерируемые объекты в <font color='6495ED'>*Python*</font> содержат внутри себя итератор. Чтобы получить его, необходимо использовать функцию <font color='6495ED'>**iter()**</font>.
Когда мы создаём цикл <font color='6495ED'>*for*</font> по итерируемому объекту (например, списку), интерпретатор на самом деле неявно обращается к его итератору через <font color='6495ED'>**iter()**</font> и использует функцию <font color='6495ED'>*next()*</font>, чтобы получить следующий элемент. Когда цикл доходит до конца объекта, <font color='6495ED'>*for*</font> отлавливает исключение <font color='red'>*StopIteration*</font> и прекращает свою работу. Вот она — магия <font color='6495ED'>*Python*</font>, которая скрыта от наших глаз. Именно так «под капотом» работает цикл <font color='6495ED'>*for*</font>, и теперь вы обладаете этим «тайным знанием».
В <font color='6495ED'>*Python*</font> можно создавать свои итераторы и прописывать в них правила выдачи элементов. Например, можно прописать, что выдавать элементы можно только по субботам, или изображения для обучения модели выдаются не по одному, а по 10 штук. Для этого создаются специальные классы с определёнными возможностями (о классах мы поговорим в модуле по ООП). Альтернативным вариантом такого подхода являются генераторы.

Наверняка у вас возник вопрос: а как это связано с функциями? Действительно, всё изложенное выше не даёт нам новых суперспособностей во владении функциями. Мы могли пользоваться циклом for и без знания того, как он устроен внутри. Но понимание работы итераторов приближает нас к пониманию более важной темы — генераторов. Всему своё время — совсем скоро мы узнаем ответ на наш вопрос, а пока предлагаем давайте закрепим пройденный материал 

### <font color='6495ED'>ГЕНЕРАТОРЫ</font>

Вы наверняка слышали словосочетание «генератор случайных чисел». Это какой-то «чёрный ящик», который по запросу пользователя выдаёт случайное число. В этом разделе мы постараемся понять, как устроены генераторы.

Генераторы случайных чисел в <font color='6495ED'>*Python*</font> — это один из типов генераторов.

Генераторы — это объекты, которые при создании не вычисляют своё содержимое, но генерируют его в процессе работы. Они выдают своё содержимое по определённым правилам только по запросу разработчика.

Правила получения объекта задаёт сам программист. Например, в генераторе можно прочитать изображение, а затем уменьшить его размер, перед тем как выдать. Вы сами — творцы логики работы генератора, и благодаря этому генераторы могут стать очень мощным инструментом для работы с огромными наборами данных.

Генератор объявляется как обычная функция. В теле функции описывается то, как будут генерироваться выдаваемые генератором объекты. Синтаксис имеет одно очень важное отличие: вместо ключевого слова <font color='6495ED'>*return*</font> в генераторе используется <font color='6495ED'>*yield*</font>. 

Проще всего рассмотреть работу генератора на примере.

Пусть у нас есть некоторая сумма денег, и мы хотим сделать неограниченный по времени вклад под определённый процент. Например, если мы внесём 10 000 рублей под 5 % годовых, то на следующий год сумма вклада будет составлять: 10 000 * 1.05 = 10 500. Ещё через год: 10 500 * 1.05 = 11 025. И так далее.

Давайте напишем генератор денег — такой генератор, который будет принимать в качестве начальных значений сумму денег и процент по вкладу, а затем с каждым вызовом генератора будет возвращать состояние вклада через год с учётом капитализации.

Код будет иметь следующий вид:

In [6]:
# Объявляем функцию для расчёта суммы вклада
def deposit(money, interest):
    # Процент по вкладу преобразуем во множитель:
    # делим процент на 100 и прибавляем 1
    interest = interest/100 + 1
    # Создаем бесконечный цикл
    while True:
        # Сумма вклада через год — это
        # текущая сумма, умноженная на коэффициент и
        # округлённая до двух знаков после запятой
        money = round(interest * money, 2)
        # Выдаём полученную сумму вклада
        yield money

Итак, мы создали функцию <font color='6495ED'>*deposit()*</font> с аргументами <font color='6495ED'>*money*</font> и <font color='6495ED'>*interest*</font>. В теле функции мы вычисляем коэффициент (1 + процент/100). Затем запускаем бесконечный цикл, в теле которого переопределяем переменную <font color='6495ED'>*money*</font> — умножаем текущую сумму на коэффициент и округляем результат до второго знака. После этого мы с помощью оператора <font color='6495ED'>*yield*</font> выводим текущую сумму на счёте.

Последняя строка нашей функции является ключевой. Именно оператор <font color='6495ED'>*yield*</font> (а не <font color='6495ED'>*return*</font>) превращает функцию в генератор.

Давайте убедимся в этом. Вызовем функцию <font color='6495ED'>*deposit()*</font> с параметрами 1000 и 5 и попробуем вывести результат на экран:

In [ ]:
print(deposit(1000, 5))
#Будет выведено
#generator object deposit at 0x0000025DAA415270

<generator object deposit at 0x000001987CC81310>


Наша функция вместо обещанных денег вернула объект типа <font color='6495ED'>*generator*</font> с именем <font color='6495ED'>*deposit*</font>. Давайте разбираться, почему функция начала возвращать нечто непонятное.

Если вместо <font color='6495ED'>*return*</font> в функции используется <font color='6495ED'>*yield*</font>, то при вызове функция возвращает объект-генератор. Сама функция при этом не выполняется, пока мы не обратимся к генератору с помощью уже знакомой нам функции <font color='6495ED'>*next()*</font>.

Таким образом, генераторы действуют по принципу: «Решаем задачи по мере их поступления, пока задач нет — отдыхаем».

Чем же так важен <font color='6495ED'>*yield*</font> и чем он отличается от <font color='6495ED'>*return*</font>?

Оператор <font color='6495ED'>*yield*</font> позволяет возвращать результат в основной блок кода и «замораживать» выполнение функции. Благодаря ему интерпретатор запоминает место, на котором он завершил работу с генератором, и возвращается на то же место при повторном обращении к генератору с помощью <font color='6495ED'>*next()*</font>.

В случае с <font color='6495ED'>*return*</font> выполнение функции завершилось бы после первой итерации <font color='6495ED'>*while*</font>, и было возвращено значение переменной.

Попробуйте сами: замените <font color='6495ED'>*yield*</font> на <font color='6495ED'>*return*</font> и вызовите функцию.

Давайте создадим «новый вклад в банке» на 1 000 рублей со ставкой 5% годовых и занесём генератор в переменную bank:

In [ ]:
bank = deposit(1000, 5)

Воспользуемся функцией <font color='6495ED'>*next()*</font> три раза, чтобы узнать, сколько денег будет на вкладе через три года:

In [ ]:
print(next(bank)) # Запускаем генератор bank в первый раз
print(next(bank)) # Запускаем генератор bank во второй раз
print(next(bank)) # Запускаем генератор bank в третий раз
# Будет напечатано:
# 1050.0
# 1102.5
# 1157.62

1050.0
1102.5
1157.62


Схема работы нашего кода и её описание представлены ниже:

![](dst3-u1-md5.2_4_3.png)

Мы создали объект-генератор <font color='6495ED'>*bank*</font>, вызвав функцию <font color='6495ED'>*deposit()*</font>. Далее мы трижды применили к генератору функцию <font color='6495ED'>*next()*</font> для того, чтобы получить следующий элемент из генератора.

- Первый вызов <font color='6495ED'>*next()*</font>: тело функции выполняется — рассчитывается переменная <font color='6495ED'>*interest*</font>. После этого начинается первая итерация цикла <font color='6495ED'>*while*</font> внутри функции <font color='6495ED'>*deposit()*</font>; оператор <font color='6495ED'>*yield*</font> вернул значение локальной переменной <font color='6495ED'>*money (round(1000 * 1.05, 2) = 1050)*</font> из функции в основной блок кода. При этом сама функция <font color='6495ED'>*deposit()*</font> не прекратила свою работу, а просто была «заморожена» до следующего вызова <font color='6495ED'>*next()*</font>.
- Второй вызов <font color='6495ED'>*next()*</font>: выполнилась вторая итерация цикла <font color='6495ED'>*while*</font>, и оператор <font color='6495ED'>*yield*</font> вернул обновлённое значение <font color='6495ED'>*money (round(1050 * 1.05, 2) = 1102.5)*</font> из функции в основной блок кода. Работа функции вновь была «заморожена», но не прекращена. Ждём следующего вызова <font color='6495ED'>*next()*</font>.
- Третий вызов <font color='6495ED'>*next()*</font>: цикл <font color='6495ED'>*while*</font> повторяется в третий раз, и оператор <font color='6495ED'>*yield*</font> возвращает новое значение переменной <font color='6495ED'>*money (round(1102.5 * 1.05, 2) = 1157.62*</font>) из функции. Снова переходим в режим ожидания следующего вызова <font color='6495ED'>*next()*</font>.

Мы разработали функцию-генератор, с помощью которой можно получать значения суммы вклада, если его срок является бесконечным. Наш генератор работает на цикле <font color='6495ED'>*while*</font> без условия выхода, поэтому может возвращать объекты бесконечно много раз.

Давайте добавим генератор вклада, который автоматически закрывается через заданное число лет. Для этого объявим функцию <font color='6495ED'>*deposit_years()*</font> с тремя аргументами: money — начальная сумма на счёте, <font color='6495ED'>*interest*</font> — процент по вкладу и <font color='6495ED'>*years*</font> — срок вклада (в годах). В отличие от прошлой функции, мы будем обновлять значение <font color='6495ED'>*money*</font> не в бесконечном цикле, а в цикле <font color='6495ED'>*for*</font> по диапазону от <font color='6495ED'>*0*</font> до <font color='6495ED'>*years*</font>. Не забудем выводить переменную money из функции с помощью <font color='6495ED'>*yield*</font>, тем самым создавая генератор:

In [ ]:
def deposit_years(money, interest, years):
    interest = interest/100 + 1
    # Вместо while используем цикл for с range
    for year in range(years):
        money = round(interest * money, 2)
        # Выдаём полученную сумму вклада
        yield money

Создадим новый вклад на два года на сумму 1 500 с 3% годовых:

In [ ]:
bank2 = deposit_years(1500, 3, 2)

Попробуем теперь три раза подряд узнать сумму денег на вкладе:

In [ ]:
print(next(bank2))
print(next(bank2))
#!print(next(bank2))
# Будет напечатано:
# 1545.0
# 1591.35
# StopIteration:

1545.0
1591.35


В первые два раза были напечатаны суммы денег, а на третий раз возникла традиционная для итератора ошибка <font color='red'>*StopIteration*</font>. Откуда она взялась?

Дело в том, что цикл <font color='6495ED'>*for*</font> в генераторе завершился после двух итераций, интерпретатор <font color='6495ED'>*Python*</font> не встретил оператор <font color='6495ED'>*yield*</font>, на котором можно было бы заморозить выполнение генератора, и окончательно прекратил работу генератора. 

Генераторы — это специальный вид итераторов. Это значит, что генераторы можно обрабатывать в цикле <font color='6495ED'>*for*</font>.

Давайте положим в наш банк 10000 с 10 % ставкой на 3 года. Затем выведем суммы вклада на каждый год на экран:

In [ ]:
bank3 = deposit_years(10000, 10, 3)
for money in bank3:
    print(money)
# Будет напечатано:
# 11000.0
# 12100.0
# 13310.0

11000.0
12100.0
13310.0


*Напомним, что на каждой итерации цикл <font color='6495ED'>for</font> невидимо для нас вызывает функцию <font color='6495ED'>next()</font>, чтобы получить следующий элемент из итератора (в нашем примере правильнее сказать генератора). Когда возникает ошибка <font color='red'>*StopIteration*</font>, цикл <font color='6495ED'>for</font> её отлавливает и прекращает итерации.*

Иногда нам хочется получить сразу все сгенерированные значения и поместить их в список, кортеж или ещё куда-либо.  

Все значения, которые генерирует функция <font color='6495ED'>*deposit_years()*</font>, можно завернуть в список, просто обернув результат её работы в функцию <font color='6495ED'>*list()*</font>:

In [ ]:
bank3 = deposit_years(10000, 10, 3)
sums = list(bank3)
print(sums)
# Будет напечатано:
# [11000.0, 12100.0, 13310.0]

[11000.0, 12100.0, 13310.0]


Когда мы обернули генератор в функцию <font color='6495ED'>*list()*</font>, мы заставили его выдать нам все свои значения и занесли их в список.

А вот с бесконечными генераторами так лучше не делать — возникнет бесконечный цикл, поскольку генератор не сможет остановить своё выполнение. ⛔ <font color='red'>**Код ниже не нужно запускать!**</font>

In [8]:

#! bank3 = deposit(10000, 10)
 
#! sums = list(bank3)
#! print(sums)
#! Возникнет бесконечный цикл! Не запускать!

Давайте на основе вышеизложенного ответим на два вопроса:

- В чём отличие генераторов от итераторов?

- В чём преимущество использования генераторов?

Генератор — это «ленивый» итератор. Отличие от обычного итератора состоит в том, что обычный итератор хранит все свои элементы в памяти, а генератор — нет. При создании генератор не выполняет код внутри себя (внутри функции). Он хранит в памяти только последний вычисленный элемент и правило перехода к следующему.

Вычисление следующего значения происходит лишь при выполнении метода <font color='6495ED'>*next()*</font>. Предыдущее значение при этом теряется. Например, в наших примерах мы генерировали бесконечную и конечную последовательность, управляя значением только одной переменной — <font color='6495ED'>*money*</font>. Мы не сохраняли предыдущие значения нашего депозита — только выдавали текущее.

***Таким образом мы получаем экономию расходуемой памяти. Это главное преимущество генераторов при работе с большим объёмом данных.***

Если проводить аналогию, то итераторы — это магазин готовой еды, где хранятся все блюда из меню и продавец просто выдаёт их вам с витрины. Генераторы — это рестораны, которые готовят блюдо только по заказу. Как только блюдо было приготовлено, оно выдаётся клиенту, и повар ждёт следующего заказа.

### <font color='6495ED'>СОКРАЩЁННАЯ ЗАПИСЬ ГЕНЕРАТОРОВ</font>

Ранее мы научились создавать генераторы в длинном виде. Синтаксис по их созданию очень похож на синтаксис для функций. Однако в <font color='6495ED'>*Python*</font> существует способ для создания простых генераторов в одну строку.

Давайте рассмотрим пример.

Представим ситуацию, мы забыли таблицу квадратов чисел и, чтобы её вспомнить, мы хотим сгенерировать квадраты натуральных чисел от 1 до 10. Напишем для этого генератор <font color='6495ED'>*square_gen*</font>:

In [3]:
# Объявляем генератор square_gen
def square_gen():
    # Создаём цикл для последовательности от 1 до 10
    for x in range(1, 11):
        # Выдаём квадрат числа
        yield x ** 2
gen = square_gen()
print(type(gen))
# Будет выведено:
# <class 'generator'>

<class 'generator'>


Отлично, генератор создан. Можно даже обернуть сгенерированные числа в список:

In [4]:
square_list = list(gen)
print(square_list)
# Будет выведено
# [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


Но не кажется ли вам, что для такой простой задачи мы написали слишком много кода? Оказывается, можно решить её одной строкой кода с помощью сокращённой формы генератора.

Для сокращённой формы генератора используются круглые скобки. В них указывается элемент, который должен быть возвращён из генератора с помощью <font color='6495ED'>*yield*</font>, а затем записывается цикл.

Ниже представлена схема перехода от общей формы генератора к сокращённой:

![](dst3-u1-md5.2_4_4.png)

Например, сгенерируем последовательность из квадратов чисел от 1 до 10. Результат занесём в переменную <font color='6495ED'>*squares_generator*</font> и убедимся, что тип данных <font color='6495ED'>*squares_generator*</font> — это <font color='6495ED'>*generator*</font>:

In [5]:
# Создаём сокращённую форму генератора квадратов
squares_generator = (x**2 for x in range(1, 11))
print(type(squares_generator))
print(squares_generator)
# Будет напечатано:
# <class 'generator'>
# <generator object <genexpr> at 0x7fc9aeab61d0>

<class 'generator'>
<generator object <genexpr> at 0x0000019127CB8A50>


Как видите, объект <font color='6495ED'>*squares_generator*</font> действительно является генератором, с которым можно работать точно так же, как с итераторами и генераторами. Вы можете использовать в нём функцию <font color='6495ED'>*next()*</font>, чтобы последовательно получить элементы или пройтись по ним в цикле <font color='6495ED'>*for*</font>. Попробуйте сделать это самостоятельно.

Можно получить список из сгенерированных элементов с помощью функции <font color='6495ED'>*list()*</font>:

In [6]:
# Создаём сокращённую форму генератора квадратов
squares_generator = (x**2 for x in range(1, 11))
# Оборачиваем сгенерированные элементы в список:
squares_list = list(squares_generator)
print(squares_list)
# Будет напечатано:
# [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


Кстати, если после создания переменной <font color='6495ED'>*squares_generator*</font> просто окружить её квадратными скобками, чуда не случится: будет просто создан список, в котором в качестве элемента будет храниться ссылка на генератор:

In [7]:
# Создаём сокращённую форму генератора квадратов
squares_generator = (x**2 for x in range(1, 11))
# Создаём список с одним элементом (генератором)
squares_list = [squares_generator]
print(squares_list)
# Будет напечатано:
# [<generator object <genexpr> at 0x7fc9aeab6bd0>]

[<generator object <genexpr> at 0x0000019127CB8970>]


По аналогии с <font color='6495ED'>*list()*</font> из генератора можно получить <font color='6495ED'>*tuple()*</font> — кортеж и <font color='6495ED'>*set()*</font> — множество:

In [8]:
# Создаём сокращённую форму генератора квадратов
squares_generator = (x**2 for x in range(1, 11))
# Оборачиваем сгенерированные значения в кортеж
squares_tuple = tuple(squares_generator)
print(squares_tuple)
# Будет напечатано:
# (1, 4, 9, 16, 25, 36, 49, 64, 81, 100)

(1, 4, 9, 16, 25, 36, 49, 64, 81, 100)


В однострочных генераторах можно также использовать условия. В сокращённой записи условия ставятся сразу после цикла без всяких двоеточий.

Например, посчитаем кубы первых 15 натуральных чисел, которые делятся на 3. Посмотрим на схему соответствия полного генератора и его сокращённой записи в таком случае:

![](dst3-u1-md5.2_4_5.png)

In [9]:
# Создаём генератор кубов чисел, которые делятся на 3
triple_cubes_generator = (x**3 for x in range(1,16) if x % 3 == 0)
print(type(triple_cubes_generator))
# Оборачиваем сгенерированные значения в список
print(list(triple_cubes_generator))
# Будет напечатано:
# <class 'generator'>
# [27, 216, 729, 1728, 3375]

<class 'generator'>
[27, 216, 729, 1728, 3375]


**Важное замечание:** можно генерировать списки сразу, без необходимости применения функции <font color='6495ED'>*list()*</font> к объекту-генератору. Достаточно написать выражение для генератора в квадратных скобках, а не в круглых. Такая конструкция называется генератором списка:

In [11]:
# Создаём генератор списка кубов чисел, которые делятся на 3
triple_cubes_list = [x**3 for x in range(1,16) if x % 3 == 0]
print(type(triple_cubes_list))
print(triple_cubes_list)
# Будет напечатано:
# <class 'list'>
# [27, 216, 729, 1728, 3375]

<class 'list'>
[27, 216, 729, 1728, 3375]


То же самое работает и для множества — достаточно написать фигурные скобки вместо круглых. Такая запись будет называться генератором множества:

In [12]:
# Создаём генератор множества кубов чисел, которые делятся на 3
triple_cubes_set = {x**3 for x in range(1,16) if x % 3 == 0}
print(type(triple_cubes_set))
print(triple_cubes_set)
# Будет напечатано:
# <class 'set'>
# {1728, 3375, 216, 729, 27}

<class 'set'>
{1728, 3375, 216, 729, 27}


А вот кортеж так получить не выйдет, поскольку круглые скобки используются для создания генераторов. Но и эту проблему можно решить — достаточно окружить выражение для генератора функцией <font color='6495ED'>*tuple()*</font>. Такая запись будет называться генератором кортежей:

In [13]:
triple_cubes_tuple = tuple(x**3 for x in range(1,16) if x % 3 == 0)
print(type(triple_cubes_tuple))
print(triple_cubes_tuple)
# Будет напечатано:
# <class 'tuple'>
# (27, 216, 729, 1728, 3375)

<class 'tuple'>
(27, 216, 729, 1728, 3375)


## <font color='6495ED'>*BONUS 6. map(), filter(), zip(), reduce() functions</font>

## <font color='6495ED'>*BONUS 7. Decorators</font>